### DISCUSSION TITLE
Running `cargo clippy` in Makefile tries to install `cargo-clippy` crate

### DISCUSSION THREAD

**Discussion by Marcono1234** [sagiegurari/cargo-make#1077] [Q&A]
Hello,
I have a repository with [`rust-toolchain.toml` file](https://github.com/Marcono1234/struson/blob/main/rust-toolchain.toml) and upgraded to the toolchain 1.78.0. The problem is that builds with [my cargo-make Makefile](https://github.com/Marcono1234/struson/blob/edf89006037963ecdc0787aa2c096be902e3be77/Makefile.toml#L77-L79) fail on CI now:
> ...
> [cargo-make] INFO - Running Task: clippy
> [cargo-make] INFO - Execute Command: "cargo" "install" "cargo-clippy"
>     Updating crates.io index
> error: could not find `cargo-clippy` in registry `crates-io` with version `*`
> [cargo-make] ERROR - Error while executing command, exit code: 101
> [cargo-make] WARN - Build Failed.

([GitHub actions log](https://github.com/Marcono1234/struson/actions/runs/8945173137/job/24576407642#step:5:1638))

For whatever reason it does not use the official `cargo clippy` but tries to install a crate called `cargo-clippy`. Do you have any idea what is causing this? Or whether cargo-make or `cargo` is performing this incorrect `cargo install`?

Sorry if the error is completely obvious, or I am misunderstanding cargo-make or cargo behavior.

I thought if I adjust `rust-toolchain.toml` and then run any `cargo` command it would automatically download the needed toolchain and components, and so far it seems that worked fine for me in the past for other toolchain versions. And also for 1.78.0 it looks like it worked locally for me without any issues (but maybe I had executed another `cargo` command before running `cargo make`).

- Am I misunderstanding how `rust-toolchain.toml` works or how cargo-make interacts with it?
- Am I misunderstanding the `Makefile.toml` syntax, or is [my `Makefile.toml`](https://github.com/Marcono1234/struson/blob/main/Makefile.toml) error-prone or ambiguous?

**Comment by Marcono1234** [comment]
My GitHub workflow was previously using cargo-make 0.37.2; I have now updated to cargo-make 0.37.11 and that seems to have solved it[^1].

Was there a change in cargo-make in the latest versions which has fixed this? Or is my Makefile or setup still faulty, and it just happens to work now for whatever reason.

[^1]: After some more testing, it looks like upgrading from cargo-make 0.37.6 to 0.37.7 fixed the GitHub workflow failure.

**Reply by sagiegurari** [reply]
i do'nt remember  a change specific to clippy

**Reply by Marcono1234** [reply]
I have tested a bit more, and it does not seem to be specific to clippy; if I reorder the tasks I get a similar issue for `cargo fmt` as well (but interestingly not for `cargo doc`):
> ...
> [cargo-make] INFO - Running Task: format-check
> [cargo-make] INFO - Execute Command: "cargo" "install" "cargo-fmt"
>     Updating crates.io index
> error: could not find `cargo-fmt` in registry `crates-io` with version `*`

For 0.37.7 your release notes say "Detect crate installation fix for new cargo list format" (and some of your commits for that version say "cargo plugin detection fix"). Could that maybe be the reason why this was solved in that version?
Though I am not familiar with what exactly this line in the release notes refers to, respectively what has changed for `cargo` in recent versions and why this fix was needed. If possible, could you please share some details about this?

Or do you notice that something is in general wrong with my setup of using `rust-toolchain.toml` in combination with cargo-make?
